045-kafka-roller-refactoring.md

```diff
@@ -0,0 +1,100 @@
+- Tension with Cruise Control e.g. over leadership.
+
+## Motivation
+
+Refactoring the Kafka Roller is a change that needs to be well planned and executed since the KafkaRoller is a crucial and complex component of the Cluster Operator.
+
+That is why this proposal aims for:
+- Proposing the desired state of the Kafka Roller.
+- Proposing a safe implementation path that focuses on incremental small changes which lead to achieving such desired state.
+
+## Proposal
+
+### Enhancement Areas
+
+The proposal focuses on the following areas of enhancements
+
+[//]: # (TODO: Change name)  
+
+#### Observations
+
+* Kafka Roller decisions can become more accurate and better informed by having rich observations coming from different sources (e.g. Kafka admin API, Kubernetes API, metrics, etc.).
+* Creating an abstraction for the observations' collection that encapsulates different observation sources, including custom sources (e.g. a Kafka Agent). Such custom sources would help with better understanding of the Kafka server state (e.g. broker is in log recovery).
+
+#### State Classification
+
+* Kafka servers can be classified into a set of states that are determined by the collected observations (e.g. needs restart, unhealthy, stable, etc.)
+* The Kafka server states can be modeled as a state machine which reflects the Kafka Roller's understanding of the Kafka server state and determines the possible actions needed to transition the Kafka server to a healthy state.
+* The Kafka Roller can continuously collect observations, classify the Kafka server state, and reconciling Kafka servers in unhealthy states via the defined state machine transitions.
+
+#### Decoupling Observations' Collection, Classification and Pod Management
+
+A new set of abstractions can be introduced around observations' collection, classification and pod management that would decouple those three aspects for the following reasons:   
+* Breaking down the complexity of the Kafka Roller and making it easier to reason about.
+* Better testability of each aspect of the roller in isolation.
+* Easier extensibility of the Kafka Roller to support future scenarios (e.g. different sources of observations, different classification logic, etc.)
+
+### Proposed Abstractions
+
+![](./images/045-kafka-roller-refactoring-design.png)
+
+#### The Observer
+
+* An abstraction around observations' collections.
+* It encapsulates the logic of collecting observations per Kafka server from different sources (e.g. Kafka admin API, Kubernetes API, metrics, etc.).
+
+#### The Classification
+
+* Responsible for mapping observations to states.
+* Does not understand state transitions.
+
+#### The Processor
+
+* An abstraction around Kafka server management. 
+* Responsible for pod deletion and reconfiguration
+* Does not understand state transitions.
+
+### State Machine
+
+The following diagram illustrates the proposed Kafka server states and possible transitions
```

Comment:
1. You should describe the states in detail.
2. For those states that imply the operator taking an action you should describe what the action is.
3. For those states that do not imply the operator taking an action you should explain what happens in the Kafka cluster that will cause a transition. This is a bit weird, in a way, because the state machine we've defined has transitions that are not caused by the operator. They're informed based on how we _expect_ the cluster to behave. But we purposely don't want to build those into the roller as _expectations_. For example, a broker can go from `stable` (i.e. running and in the ISR) to `syncing` (i.e. running but not in the ISR) because of producer load/interbroker throttling. Those are beyond the roller's ability to actively influence, but they do change the behaviour of the roller: To move on to the next pod we want to see the broker joint the ISR at least momentarily. If it later falls out of the ISR the roller doesn't directly care: It's not going to cause the roller to try restarting the broker; but it might cause us to delay restarting another broker if it would impact topic availability.